tests/test_ilastik/test_workflows/test_HeadlessPixelClassificationWorkflow.py

```diff
@@ -0,0 +1,178 @@
+    testdir,
+    *,
+    num_distributed_workers: int = 0,
+    project: Path,
+    raw_data: Path,
+    output_filename_format: str,
+    input_axes: str = "",
+    output_format: str = "hdf5",
+    ignore_training_axistags: bool = False,
+):
+    assert project.exists()
+    assert raw_data.parent.exists()
+
+    ilastik_dot_py = Path(__file__).parent.parent.parent.parent / "ilastik.py"
+    subprocess_args = [
+        "python",
+        str(ilastik_dot_py),
+        "--headless",
+        "--project=" + str(project),
+        "--raw-data=" + str(raw_data),
+        "--output_filename_format=" + str(output_filename_format),
+        "--output_format=" + output_format,
+    ]
+
+    if input_axes:
+        subprocess_args.append("--input-axes=" + input_axes)
+
+    if ignore_training_axistags:
+        subprocess_args.append("--ignore_training_axistags")
+
+    if num_distributed_workers:
+        os.environ["OMPI_ALLOW_RUN_AS_ROOT"] = "1"
+        os.environ["OMPI_ALLOW_RUN_AS_ROOT_CONFIRM"] = "1"
+        subprocess_args = ["mpiexec", "-n", str(num_distributed_workers)] + subprocess_args + ["--distributed"]
+
+    result = testdir.run(*subprocess_args)
+    if result.ret != 0:
+        raise FailedHeadlessExecutionException(
+            "===STDOUT===\n\n" + result.stdout.str() + "\n\n===STDERR===\n\n" + result.stderr.str()
+        )
+
+
+def test_headless_2d3c_with_same_raw_data_axis(testdir, pixel_classification_ilp_2d3c: Path, tmp_path: Path):
+    raw_100x100y3c: Path = create_h5(numpy.random.rand(100, 100, 3), axiskeys="yxc")
+    output_path = tmp_path / "out_100x100y3c.h5"
+    run_headless_pixel_classification(
+        testdir, project=pixel_classification_ilp_2d3c, raw_data=raw_100x100y3c, output_filename_format=str(output_path)
+    )
+
+
+def test_headless_2d3c_with_swizzled_raw_data_axis(testdir, pixel_classification_ilp_2d3c: Path, tmp_path: Path):
+    raw_3c100x100y: Path = create_h5(numpy.random.rand(3, 100, 100), axiskeys="cyx")
+    output_path = tmp_path / "out_3c100x100y.h5"
+
+    # default behavior is to try to apply training axistags to the batch data, and therefore fail because raw data's
+    # dimensions do not match that of the training data
+    with pytest.raises(FailedHeadlessExecutionException):
+        run_headless_pixel_classification(
+            testdir,
+            project=pixel_classification_ilp_2d3c,
+            raw_data=raw_3c100x100y,
+            output_filename_format=str(output_path),
+        )
+
+    # forcing correct input axes should pass
+    run_headless_pixel_classification(
+        testdir,
+        project=pixel_classification_ilp_2d3c,
+        raw_data=raw_3c100x100y,
+        output_filename_format=str(output_path),
+        input_axes="cyx",
+    )
+
+    # alternatively, since the generated h5 data has the axistags property, we can ignore training data and use that
+    # instead, by using the '--ignore_training_axistags' flag
+    run_headless_pixel_classification(
```

Comment:
hang on. Do I understand correctly, that the default behavior now is:

* try to force the training data axistags (probably the ones of the last lane), even if something like proper axistags comes with the data
* if `--ignore-training-flags` is passed, only then the axistags in the file are used.

This seems wrong. Shou;dn't the logic (`DataSelectionGui.guess_axistags_for`) be the same in headless mode as in gui mode? (only apply lane axistags if those were changed?)

Reply:
Hmmm I think I agree with you that (even though I don't like it) it would be more consistent to only apply the training axistags onto the new data if the training axistags were changed from the default. Still, we run into the problem that when running headless, the input data may not be available any more. In that case, I don't think there is a way to tell if the axistags have been forced or not. We could save that into the `.ilp` files, but there would be no backwards compatibility.

Also, forcing the  axistags from the last lane is the legacy behavior anyway (create project, permute axis, run headless, check output: it is also permuted), so it might be a good idea to keep that behavior as it is.

Reply:
allright, let's keep it this way then.